source/common/router/scoped_rds.cc

```diff
@@ -121,19 +121,98 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
   });
 }
 
+// Constructor for RdsRouteConfigProviderHelper when scope is eager loading.
+// Initialize RdsRouteConfigProvider by default.
 ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
     ScopedRdsConfigSubscription& parent, std::string scope_name,
     envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
     Init::Manager& init_manager)
-    : parent_(parent), scope_name_(scope_name),
-      route_provider_(std::dynamic_pointer_cast<RdsRouteConfigProviderImpl>(
-          parent_.route_config_provider_manager_.createRdsRouteConfigProvider(
-              rds, parent_.factory_context_, parent_.stat_prefix_, init_manager))),
+    : parent_(parent), scope_name_(scope_name), on_demand_(false) {
+  initRdsConfigProvider(rds, init_manager);
+}
+
+// Constructor for RdsRouteConfigProviderHelper when scope is on demand.
+// Leave the RdsRouteConfigProvider uninitialized.
+ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
+    ScopedRdsConfigSubscription& parent, std::string scope_name)
+    : parent_(parent), scope_name_(scope_name), on_demand_(true) {
+  parent_.stats_.on_demand_scopes_.inc();
+}
+
+// When on demand callback is received from main thread, there are 4 cases.
+// 1. Scope is not found, post a scope not found callback back to worker thread.
+// 2. Scope is found but route provider has not been initialized, create route provider.
+// 3. After route provider has been initialized, if RouteConfiguration has been fetched,
```

Comment:
I was thinking a bit about the dance that is done between worker/main threads. Imagine we have a situation where we keep taking a miss on SRDS. Is it possible to DoS the main thread by having let's say 100 worker threads continually posting requests to the main thread? This could happen even accidentally, as we scale worker threads in an SRDS Envoy. 

I'm not sure if this needs any radical redesign, but it would be good to be able to determine super early, ideally on the worker thread, whether there will be any benefit in posting to the main thread, e.g. by being able to determine if there is a valid (potential) route provider or the like.

@stevenzzzz @chaoqin-li1123 WDYT?

Reply:
We can check whether the scope exists in worker threads. On a srds update, if a scope is on demand, we still propagate its key and name to scoped_config_impl in worker thread. When we post a on demand update from a worker thread, we are quite confident that the scope exists. Although the scope may be overwritten by another update later when the main thread execute the callback.

Reply:
The computeScopeKey function will return a nullptr when scope doesn't exist, in that case callback is not posted to main thread.

Reply:
This makes sense. Can you document this somewhere in code comments? Thanks.

Reply:
Will do.